lang/java/compiler/src/main/java/org/apache/avro/compiler/idl/SchemaResolver.java

```diff
@@ -0,0 +1,224 @@
+      to.addProp(entry.getKey(), entry.getValue());
+    }
+  }
+
+  static Schema resolve(final Schema schema, final Protocol protocol, final Map<String, Schema> processed) {
+    final String fullName = schema.getFullName();
+    if (fullName != null && processed.containsKey(fullName)) {
+      return processed.get(schema.getFullName());
+    } else if (isUnresolvedSchema(schema)) {
+      final String unresolvedSchemaName = getUnresolvedSchemaName(schema);
+      Schema type = protocol.getType(unresolvedSchemaName);
+      if (type == null) {
+        throw new IllegalArgumentException("Cannot resolve " + unresolvedSchemaName);
+      }
+      return resolve(type, protocol, processed);
+    } else {
+      switch (schema.getType()) {
+        case RECORD:
+          Schema createRecord = Schema.createRecord(schema.getName(), schema.getDoc(), schema.getNamespace(),
+                  schema.isError());
+          processed.put(schema.getFullName(), createRecord);
+          final List<Schema.Field> currFields = schema.getFields();
+          List<Schema.Field> newFields = new ArrayList<Schema.Field>(currFields.size());
+          for (Schema.Field field : currFields) {
+            Schema.Field nf = new Schema.Field(field.name(), resolve(field.schema(), protocol, processed),
+                    field.doc(), field.defaultVal(), field.order());
+            for (String alias : field.aliases()) {
+              nf.addAlias(alias);
+            }
+            newFields.add(nf);
+          }
+          createRecord.setFields(newFields);
+          final LogicalType lt = schema.getLogicalType();
+          if (lt != null) {
+            lt.addToSchema(createRecord);
+          }
+          copyProps(schema, createRecord);
+          return createRecord;
+        case MAP:
+          Schema result = Schema.createMap(resolve(schema.getValueType(), protocol, processed));
+          copyProps(schema, result);
+          return result;
+        case ARRAY:
+          Schema aresult = Schema.createArray(resolve(schema.getElementType(), protocol, processed));
+          copyProps(schema, aresult);
+          return aresult;
+        case UNION:
+          final List<Schema> uTypes = schema.getTypes();
+          List<Schema> newTypes = new ArrayList<Schema>(uTypes.size());
+          for (Schema s : uTypes) {
+            newTypes.add(resolve(s, protocol, processed));
+          }
+          Schema bresult = Schema.createUnion(newTypes);
+          copyProps(schema, bresult);
+          return bresult;
+        case ENUM:
+        case FIXED:
+        case STRING:
+        case BYTES:
+        case INT:
+        case LONG:
+        case FLOAT:
+        case DOUBLE:
+        case BOOLEAN:
+        case NULL:
+          return schema;
+        default:
+          throw new RuntimeException("Unknown type: " + schema);
+      }
+    }
+  }
+
+  public static Schema intern(final Schema schema, final Map<String, Schema> processed) {
```

Comment:
Agree the method name is not optimal,

Let me remember what I as trying to do here. (here is proof that writing comments is not a waste of time, and I already regret not doing it :-) )

Here is what I was trying to do:

When resolving the "unresolved" schemas instead of mutating the objects that reference the unresolved schema I create new instances instead of mutating them (out of principle).

hope that it makes sense.